src/gen9_avc_encoder.c

```diff
@@ -0,0 +1,6031 @@
+    }
+    ret_kernel->size = next_krnoffset - (pcurr_header->kernel_start_pointer << 6);
+
+    return true;
+}
+static void
+gen9_free_surfaces_avc(void **data)
+{
+    struct gen9_surface_avc *avc_surface;
+
+    if (!data || !*data)
+        return;
+
+    avc_surface = *data;
+
+    if (avc_surface->scaled_4x_surface_obj) {
+        i965_DestroySurfaces(avc_surface->ctx, &avc_surface->scaled_4x_surface_id, 1);
+        avc_surface->scaled_4x_surface_id = VA_INVALID_SURFACE;
+        avc_surface->scaled_4x_surface_obj = NULL;
+    }
+
+    if (avc_surface->scaled_16x_surface_obj) {
+        i965_DestroySurfaces(avc_surface->ctx, &avc_surface->scaled_16x_surface_id, 1);
+        avc_surface->scaled_16x_surface_id = VA_INVALID_SURFACE;
+        avc_surface->scaled_16x_surface_obj = NULL;
+    }
+
+    if (avc_surface->scaled_32x_surface_obj) {
+        i965_DestroySurfaces(avc_surface->ctx, &avc_surface->scaled_32x_surface_id, 1);
+        avc_surface->scaled_32x_surface_id = VA_INVALID_SURFACE;
+        avc_surface->scaled_32x_surface_obj = NULL;
+    }
+
+    i965_free_gpe_resource(&avc_surface->res_mb_code_surface);
+    i965_free_gpe_resource(&avc_surface->res_mv_data_surface);
+    i965_free_gpe_resource(&avc_surface->res_ref_pic_select_surface);
+
+    dri_bo_unreference(avc_surface->dmv_top);
+    avc_surface->dmv_top = NULL;
+    dri_bo_unreference(avc_surface->dmv_bottom);
+    avc_surface->dmv_bottom = NULL;
+
+    free(avc_surface);
+
+    *data = NULL;
+
+    return;
+}
+
+static VAStatus
+gen9_avc_init_check_surfaces(VADriverContextP ctx,
+                             struct object_surface *obj_surface,
+                             struct intel_encoder_context *encoder_context,
+                             struct avc_surface_param *surface_param)
+{
+    struct i965_driver_data *i965 = i965_driver_data(ctx);
+    struct encoder_vme_mfc_context * vme_context = (struct encoder_vme_mfc_context *)encoder_context->vme_context;
+    struct avc_enc_state * avc_state = (struct avc_enc_state * )vme_context->private_enc_state;
+    struct generic_enc_codec_state * generic_state = (struct generic_enc_codec_state * )vme_context->generic_enc_state;
+
+    struct gen9_surface_avc *avc_surface;
+    int downscaled_width_4x, downscaled_height_4x;
+    int downscaled_width_16x, downscaled_height_16x;
+    int downscaled_width_32x, downscaled_height_32x;
+    int size = 0;
+    unsigned int frame_width_in_mbs = ALIGN(surface_param->frame_width,16) / 16;
+    unsigned int frame_height_in_mbs = ALIGN(surface_param->frame_height,16) / 16;
+    unsigned int frame_mb_nums = frame_width_in_mbs * frame_height_in_mbs;
+    int allocate_flag = 1;
+    int width,height;
+
+    if (!obj_surface || !obj_surface->bo)
+        return VA_STATUS_ERROR_INVALID_SURFACE;
+
+    if (obj_surface->private_data &&
+        obj_surface->free_private_data != gen9_free_surfaces_avc) {
+        obj_surface->free_private_data(&obj_surface->private_data);
```

Comment:
Yes, we should fix private data usage in another patch. What I mean is not to release private data here because releasing private data will result in other issues.  BTW 
fhvwy has created https://github.com/01org/intel-vaapi-driver/issues/42 to track this issue